### PR TITLE
ESQL: Create `Block.Ref`

### PR DESCRIPTION
**Pull request by nik9000**
This creates `Block.Ref`, a reference to a `Block` which may or may not be part of a `Page`. `Block.Ref` is `Releasable` and closing it is a noop if the `Block` is part of a `Page`, but if it is "free floating" then closing the `Block.Ref` will close the block.

It also modified `ExpressionEvaluator` to return a `Block.Ref` instead of a `Block` - so you tend to work with `ExpressionEvaluator`s like this:
```
try (Block.Ref ref = eval.eval(page)) {
  return ref.block().doStuff();
}
```

This should make it *much* easier to release the memory from `Block`s built by `ExpressionEvaluator`s.

This change is mostly mechanical, introducing the new signature for `ExpressionEvaluator`. In a follow up change I'll modify the tests to make sure we're correctly using it to close pages.

I did think about changing `ExpressionEvaluator` to add a method telling you if the block that it returns must be closed or not. This would have been more difficult to work with, and, ultimately, limiting. Specifically, it is possible for an `ExpressionEvaluator` to *sometimes* return a free floating block and other times return one that is contained in a `Page`. Imagine `mv_concat` - it returns the block it receives if the block doesn't have multivalued fields. Otherwise it concats things. If that block happens to come directly out of the `Page`, then `mv_concat` will sometimes produce free floating blocks and sometimes not.

### CODE REVIEW

**Review by ChrisHegarty — APPROVED** [review]

**Comment by nik9000** [comment]
Yeah! It's compatible with reference counting if we need to go there. And
if we don't, well, it's ok!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java [hunk 1]
```diff
@@ -208,6 +209,48 @@ interface Builder extends Releasable {
...
+     * }</pre>
+     *
+     * The {@code try} block will return the memory used by the block to the
+     * breaker if it was "free floating", but if it was attached to a {@link Page}
+     * then it'll do nothing.
+     *
+     * @param block the block referenced
+     * @param containedIn the page containing it or null, if it is "free floating".
+     */
+    record Ref(Block block, @Nullable Page containedIn) implements Releasable {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java:231]: Yeah! I had started off putting it as `EvalOperator.Result` but figured it'd be useful elsewhere so I moved it.

**Comment by nik9000** [comment]
run elasticsearch-ci/part-1

**Review by bpintea — APPROVED** [review]
Nice!

**Pull request merged**